### PR TITLE
Add support for the RISC Zero SHA-256 accelerator

### PR DESCRIPTION
**Pull request by nategraf**

### CODE REVIEW

**Comment by nategraf** [comment]
One thing I will call out in this PR is that I used `alloc` in the buffer copying when the input is not word-aligned. An alternative to this would be to avoid using `alloc`, or use it only when otherwise available, and iteratively compress a fixed size buffer instead. I chose to use `alloc` anyway because we know that the guest has it available.

File: sha2/Cargo.toml [hunk 1]
```diff
@@ -22,6 +22,10 @@ cfg-if = "1.0"
 cpufeatures = "0.2"
 sha2-asm = { version = "0.6.1", optional = true }
 
+[target.'cfg(all(target_os = "zkvm", target_arch = "riscv32"))'.dependencies]
+risc0-zkvm-platform = "0.12"
```
Comment (flaub) [inline sha2/Cargo.toml]: I wonder if there's a way to late-bind this dependency. Otherwise seems like we need to keep bumping this.

Comment (shkoo) [reply sha2/Cargo.toml]: Alternatively, maybe leave off the dependency and include the extern "C" definition for sys_sha_buffer in risc0.rs?

Comment (nategraf) [reply sha2/Cargo.toml]: When I was working on this before I did not find any nice way to do late-binding, either for extern Rust or C, dependencies with Cargo. Do y'all have any pointers that might be helpful in that regard?

**Comment by nategraf** [comment]
What are we doing with this patch? Is our goal to merge this into our fork and then open a PR again Rust-Crypto? I'd love to make sure we keep working toward that or make clear otherwise how someone can use this work.

**Comment by nategraf** [comment]
Ok, I figured out the late-binding risc0-zkvm-platform by using the `extern "C"` API and have used it to clean up this diff.

@flaub, if this looks good, I'd like to merge this branch, and add a `sha2-v0.10.6-risc0` tag as well, to us in upcoming docs on how to patch in RISC Zero accelerator support.

Note that I will use the same technique with `sys_bigint` for the patch I am finishing up for `RustCrypto/crypto-bigint`

**Review by flaub — APPROVED** [review]
LGTM

**Pull request merged**